src/ShallowWaterRosenbrock.jl

```diff
@@ -0,0 +1,252 @@
+  get_free_dof_values(h₂) .= get_free_dof_values(h₁) .+ dt .* get_free_dof_values(dh₂)
+end
+
+function new_vtk_step(Ω,file,hn,un,wn)
+  createvtk(Ω,
+            file,
+            cellfields=["hn"=>hn, "un"=>un, "wn"=>wn],
+            nsubcells=4)
+end
+
+function compute_mean_depth!(wrk, L2MM, h)
+  # compute the mean depth over the sphere, for use in the approximate Jacobian
+  mul!(wrk, L2MM, get_free_dof_values(h))
+  h_int = sum(wrk)
+  wrk  .= 1.0
+  tmp   = L2MM*wrk # create a new vector, only doing this once during initialisation
+  a_int = sum(tmp)
+  h_avg = h_int/a_int
+  println("mean depth: ", h_avg)
+  h_avg
+end
+
+function shallow_water_rosenbrock_time_stepper(model, order, degree,
+                        h₀, u₀, f₀, g, H₀,
+                        λ, dt, τ, N; leap_frog=false,
+                        write_diagnostics=true,
+                        write_diagnostics_freq=1,
+                        dump_diagnostics_on_screen=true,
+                        write_solution=false,
+                        write_solution_freq=N/10,
+                        output_dir="nswe_eq_ncells_$(num_cells(model))_order_$(order)_rosenbrock")
+
+  # Forward integration of the shallow water equations
+  Ω = Triangulation(model)
+  dΩ = Measure(Ω, degree)
+  dω = Measure(Ω, degree, ReferenceDomain())
+
+  # Setup the trial and test spaces
+  reffe_rt  = ReferenceFE(raviart_thomas, Float64, order)
+  V = FESpace(model, reffe_rt ; conformity=:HDiv)
+  U = TrialFESpace(V)
+  reffe_lgn = ReferenceFE(lagrangian, Float64, order)
+  Q = FESpace(model, reffe_lgn; conformity=:L2)
+  P = TrialFESpace(Q)
+  reffe_lgn = ReferenceFE(lagrangian, Float64, order+1)
+  S = FESpace(model, reffe_lgn; conformity=:H1)
+  R = TrialFESpace(S)
+
+  Y = MultiFieldFESpace([V, Q])
+  X = MultiFieldFESpace([U, P])
+
+  # assemble the mass matrices
+  amm(a,b) = ∫(a⋅b)dΩ
+  H1MM = assemble_matrix(amm, R, S)
+  RTMM = assemble_matrix(amm, U, V)
+  L2MM = assemble_matrix(amm, P, Q)
+  H1MMchol = lu(H1MM)
+  RTMMchol = lu(RTMM)
+  L2MMchol = lu(L2MM)
+
+  # Project the initial conditions onto the trial spaces
+  b₁(q)   = ∫(q*h₀)dΩ
+  b₂(v)   = ∫(v⋅u₀)dΩ
+  b₃(s)   = ∫(s*f₀)*dΩ
+  rhs1    = assemble_vector(b₃, S)
+  f       = FEFunction(S, copy(rhs1))
+  ldiv!(H1MMchol, get_free_dof_values(f))
+
+  # assemble the approximate MultiFieldFESpace Jacobian
+  n = get_normal_vector(model)
+
+  Amat((u,p),(v,q)) = ∫(-dt*λ*f*(v⋅⟂(u,n)))dΩ + ∫(dt*λ*g*(DIV(v)*p))dω - ∫(dt*λ*H₀*(q*DIV(u)))dω
+  Mmat((u,p),(v,q)) = ∫(u⋅v)dΩ + ∫(p*q)dΩ # block mass matrix
+  A = assemble_matrix(Amat, X,Y)
+  M = assemble_matrix(Mmat, X,Y)
+  B = M - A
+  Bchol = lu(B)
+  Mchol = lu(M)
+  # leap frog matrices, using 2×dt
+  lf = 1.0
+  if leap_frog
+    lf = 2.0
+  end
+  Blf = M - lf*A
+  Blfchol = lu(Blf)
+
+  # multifield initial condtions
+  b₄((v,q)) = b₁(q) + b₂(v)
+  rhs2    = assemble_vector(b₄, Y)
+  yn      = FEFunction(Y, copy(rhs2))
```

Comment:
Use `clone_fe_function`

Reply:
I don't think `clone_fe_function` can be used in L168 as `rhs2` is a vector

Reply:
Ok, you are right. On the other hand, I think there is no need to `copy` `rhs2`. I would rewrite these lines as:

```
rhs2    = assemble_vector(b₄, Y)
ldiv!(Mchol, rhs2)
yn  = FEFunction(Y, rhs2)
```

Reply:
Good point. I have updated this now